File: src/main/java/org/oewntk/model/LexicographicOrder.java
/*
 * Copyright (c) 2021. Bernard Bou.
 */

package org.oewntk.model;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

public class LexicographicOrder
{
	public static final Comparator<String> lowerFirst = (s1, s2) -> {
		int c = s1.compareToIgnoreCase(s2);
		if (c != 0)
		{
			return c;
		}
		// same lower-case map
		return -s1.compareTo(s2);
	};

	public static final Comparator<String> upperFirst = (s1, s2) -> {
		int c = s1.compareToIgnoreCase(s2);
		if (c != 0)
		{
			return c;
		}
		// same lower-case map
		return s1.compareTo(s2);
	};
}

File: src/main/java/org/oewntk/model/LexicographicOrder.kt
/*
 * Copyright (c) 2021. Bernard Bou.
 */
package org.oewntk.model

/**
 * Lexicographic Order
 */
object LexicographicOrder {

    /**
     * Lower-case first
     */
    val lowerFirst: Comparator<String> = Comparator { s1: String, s2: String ->
        val c = s1.compareTo(s2, ignoreCase = true)
        if (c != 0) {
            return@Comparator c
        }
        -s1.compareTo(s2)
    }

    /**
     * Upper-case first
     */
    val upperFirst: Comparator<String> = Comparator { s1: String, s2: String ->
        val c = s1.compareTo(s2, ignoreCase = true)
        if (c != 0) {
            return@Comparator c
        }
        s1.compareTo(s2)
    }
}
